### ISSUE TITLE
bug in resumable upload nginx 

### ISSUE THREAD

**Issue by slavaslava56ru** [fdintino/nginx-upload-module#123]
Hey. I have a problem. When the Internet breaks, nginx gives http code 400. I have set cleanup on it by default. When the file is downloaded, the problem arises that the upper bytes of the file are empty :( Can you tell me how to fix it? Besides removing 400 code from cleanup

**Comment by slavaslava56ru** [comment]
All that I found in debug nginx
![Снимок экрана 2019-12-08 в 11 06 23](https://user-images.githubusercontent.com/32861091/70385134-f4c76900-19ac-11ea-88ba-8b250ec3f9ce.png)

**Comment by coldrift** [comment]
I don't see any issue here. You asked the system to clean up the file and it was cleaned up as per your log.

**Comment by slavaslava56ru** [comment]
> I don't see any issue here. You asked the system to clean up the file and it was cleaned up as per your log.

Yes, but nginx throw out 400 when the file was not loaded. I do not know what to do :( 
is it possible to replace this 400 with another http code or can you tell me where it comes from. 

I reproduce the bug in the following steps
1. Starting to load part of the file
2. Turn off the internet
3. nginx return 400

**Comment by slavaslava56ru** [comment]
![Снимок экрана 2019-12-10 в 10 47 58](https://user-images.githubusercontent.com/32861091/70499281-a647d500-1b3a-11ea-9335-060fd68fc51c.png)
attach screen access log

**Comment by slavaslava56ru** [comment]
![Снимок экрана 2019-12-10 в 10 47 58](https://user-images.githubusercontent.com/32861091/70499346-c4153a00-1b3a-11ea-83a4-a4242027c868.png)
attach screen

**Comment by slavaslava56ru** [comment]
timeout after 7,5 sec

**Comment by vkholodkov** [comment]
I see HTTP/2.0 request in your log. upload module does not work with http/2.0. The reason 400 is returned is probably because of a bug in nginx.

**Comment by slavaslava56ru** [comment]
> I see HTTP/2.0 request in your log. upload module does not work with http/2.0. The reason 400 is returned is probably because of a bug in nginx.

😮  not knew. Thank you so much

**Comment by fdintino** [comment]
@vkholodkov it is not true that the module doesn’t support http 2. There are even unit tests to ensure it functions, and we’ve been using it with http 2 for over a year now, when I added that functionality

**Comment by coldrift** [comment]
@fdintino I did not know that. @slavaslava56ru if so, then the issue is somewhere in your system or nginx itself.

**Comment by vkholodkov** [comment]
@fdintino I did not know that. @slavaslava56ru if so, then the issue is somewhere in your system or nginx itself.

**Comment by slavaslava56ru** [comment]
@vkholodkov Can you tell, Where to look for a problem?

**Comment by vkholodkov** [comment]
@slavaslava56ru can I ask you where you from and what do you do with nginx and nginx upload module?

**Comment by slavaslava56ru** [comment]
![Снимок экрана 2019-12-10 в 15 29 21](https://user-images.githubusercontent.com/32861091/70521900-df477000-1b61-11ea-95b6-2cc1f9af0f2f.png)

**Comment by slavaslava56ru** [comment]
I am trying to download a file in parts. I specifically turn off the Internet to check that if the part did not upload, then everything is ok

**Comment by fdintino** [comment]
Are you uploading in the browser and the javascript File interface? If so, you would want to catch the error and retry the upload range in the event of an error. This is how we have implemented a resumable upload with offline network error handling. If you are simply ignoring the error, it makes sense that the range that errored would be filled with null bytes.

**Comment by fdintino** [comment]
It might also be helpful to diagnose the issue if you provided a gist of full debug log.

**Comment by slavaslava56ru** [comment]
Yes, through the js interface. Byte range is loaded again the same on which it broke. The problem is that nginx gives an error which clear ranges loaded before

**Comment by slavaslava56ru** [comment]
Oki, Now I will try to get a complete log

**Comment by slavaslava56ru** [comment]
![Снимок экрана 2019-12-10 в 16 07 41](https://user-images.githubusercontent.com/32861091/70524561-3e5bb380-1b67-11ea-960a-6c4176ede6c8.png)
js
![Снимок экрана 2019-12-10 в 16 07 25](https://user-images.githubusercontent.com/32861091/70524617-62b79000-1b67-11ea-9f95-ef544b58e4ae.png)
nginx

**Comment by slavaslava56ru** [comment]
![Снимок экрана 2019-12-10 в 16 07 25](https://user-images.githubusercontent.com/32861091/70524715-972b4c00-1b67-11ea-961b-66f1ae265906.png)

**Comment by slavaslava56ru** [comment]
[debug (7).log](https://github.com/fdintino/nginx-upload-module/files/3944541/debug.7.log)

**Comment by slavaslava56ru** [comment]
may be reason in this directives ```
reset_timedout_connection on;
client_body_timeout
client_header_timeout
send_timeout```

**Comment by fdintino** [comment]
Is there any chance you have the directive `upload_cleanup` somewhere in your config?

**Comment by slavaslava56ru** [comment]
upload_cleanup 400 404 500-599

**Comment by vkholodkov** [comment]
I think what happens is that you get n == NGX_ERROR at this place in ngx_http_do_read_upload_client_request_body:

```
            if (n == 0 || n == NGX_ERROR) {
                c->error = 1;
                return NGX_HTTP_BAD_REQUEST;
            }
```
The 400 status codes is not appropriate here, because the request has not been received entirely. More appropriate code is NGX_HTTP_CLIENT_CLOSED_REQUEST.

Please replace NGX_HTTP_BAD_REQUEST with NGX_HTTP_CLIENT_CLOSED_REQUEST and see if it helps your situation.

**Comment by fdintino** [comment]
Yes, @vkholodkov is right I believe. We're returning the wrong status code in that bit of code. That, combined with the `upload_cleanup 400 404 500-599` directive, are causing the file cleanup.

If you were to change the `upload_cleanup` instead to

```
upload_cleanup 404 500-599;
```

that would temporarily work around the bug, without requiring you to recompile nginx. But that's just a temporary hack. I have some time today, I'll implement the above fix and add a test for it.

**Comment by fdintino** [comment]
Hmmm.... one other thing. Are you sure you recompiled the module with debug flags? I don't see any of the nginx-upload-module debug messages in the log file you posted.

**Comment by slavaslava56ru** [comment]
Sorry, that nginx build is gone, so I can’t say for sure

**Comment by vkholodkov** [comment]
you can use nginx -V to dump the full command line that was used to run your build

**Comment by slavaslava56ru** [comment]
https://user-images.githubusercontent.com/32861091/70521900-df477000-1b61-11ea-95b6-2cc1f9af0f2f.png

**Comment by slavaslava56ru** [comment]
Today will fix?

**Comment by fdintino** [comment]
I only ever get a 408 when I try to replicate your issue. Can you try applying @vkholodkov's patch and see if it fixes things for you?

**Comment by slavaslava56ru** [comment]
@fdintino ok, now see

**Comment by slavaslava56ru** [comment]
not fixed :(

**Comment by coldrift** [comment]
If you still get 400, this means your client sends something wrong, because other places where upload module return 400 are related to request validation.

**Comment by SoNWarrioR** [comment]
> I think what happens is that you get n == NGX_ERROR at this place in ngx_http_do_read_upload_client_request_body:
> 
> ```
>             if (n == 0 || n == NGX_ERROR) {
>                 c->error = 1;
>                 return NGX_HTTP_BAD_REQUEST;
>             }
> ```
> 
> The 400 status codes is not appropriate here, because the request has not been received entirely. More appropriate code is NGX_HTTP_CLIENT_CLOSED_REQUEST.
> 
> Please replace NGX_HTTP_BAD_REQUEST with NGX_HTTP_CLIENT_CLOSED_REQUEST and see if it helps your situation.

Hello. Have same problem, and this patch not fixed it, can u help me?
